batch/batch/batch.py

```diff
@@ -147,27 +150,116 @@ def batch_record_to_dict(app, record):
         instance.adjust_free_cores_in_memory(rv['delta_cores_mcpu'])
 
 
-def job_record_to_dict(app, record, running_status=None):
-    spec = json.loads(record['spec'])
+async def _get_full_job_spec(app, record):
+    db = app['db']
+    log_store = app['log_store']
+
+    batch_id = record['batch_id']
+    job_id = record['job_id']
+    format_version = record['format_version']
+
+    if format_version == 1:
+        return record['spec']
+
+    bunch_record = await db.select_and_fetchone(
+        '''
+SELECT token, start_job_id FROM batch_bunches
+WHERE batch_id = %s AND start_job_id <= %s
+ORDER BY start_job_id DESC
+LIMIT 1;
+''',
+        (batch_id, job_id))
+    token = bunch_record['token']
```

Comment:
Hmm, I was thinking we could use (batch_id, start_job_id) for the token.  It has to be unique for each bunch, and simplifies some things (eliminates an index on bunches, etc.)

Reply:
I thought the point of the token was in case we have multiple attempts to write the same bunch. I changed the primary key to `(batch_id, start_job_id, token)` and got rid of the index on `(batch_id, start_job_id)`. Not sure if that's correct.

Reply:
The bunch should only be inserted once.  If on insert a bunch with the given start_id is already there, we should just succeed (alternatively, if a job with the given start_job_id in the bunch is already there, we should succeed -- this was the change Dan made.)

Reply:
Then why do we need the token at all?

Reply:
I was just going off of your spec in the google doc

Reply:
Oh!  I was wrong.  I don't think we do anymore.  Sounds like you agree.  Let's nuke it.

Reply:
I nuked it, but my concern is we write the spec first before checking the database whether the bunch exists already.

```
        if batch_format_version.has_full_spec_in_gcs():
            async with timer.step('write spec to gcs'):
                await spec_writer.write()

        rand_token = random.randint(0, app['n_tokens'] - 1)
        n_jobs = len(job_specs)

        async with timer.step('insert jobs'):
            @transaction(db)
            async def insert(tx):
...
```

Reply:
Ah, good point.  OK, my proposal:

Add the token back for the spec only.  You'll need it for reading the spec, but it shouldn't be part of the bunch table index.  There will only be one bunch entry in the database for each start_job_id (this is the main thing I'm advocating for).  You'll occasionally write the spec multiple times but ignore all but one.